patches/server/1055-Add-required-plugins-config-option.patch

```diff
@@ -0,0 +1,82 @@
++            this.server.getLogger().severe("The required plugin " + plugin.getName() + " got disabled. Stopping Server.");
++            this.server.shutdown();
++            return;
++        }
++        // Paper end
+         if (!plugin.isEnabled()) {
+             return;
+         }
+diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+index 1c9742ad81f04052d2c3bc18c7636f45b2fc5160..4abf65f1d423ce722dbff5c0220228d37d83c982 100644
+--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+@@ -275,6 +275,13 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
+         // CraftBukkit end
+ 
+         // Paper start
++        List<String> missingPlugins = new java.util.ArrayList<>(io.papermc.paper.configuration.GlobalConfiguration.get().misc.requiredPlugins);
++        missingPlugins.removeAll(java.util.Arrays.stream(this.server.paperPluginManager.getPlugins()).filter(org.bukkit.plugin.Plugin::isEnabled).map(org.bukkit.plugin.Plugin::getName).toList());
++        if (!missingPlugins.isEmpty()) {
++            DedicatedServer.LOGGER.error("The following required plugins aren't enabled: {}. Stopping Server.", String.join(", ", missingPlugins));
++            return false;
++        }
++
+         boolean usingProxy = org.spigotmc.SpigotConfig.bungee || io.papermc.paper.configuration.GlobalConfiguration.get().proxies.velocity.enabled;
+         String proxyFlavor = (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.velocity.enabled) ? "Velocity" : "BungeeCord";
+         String proxyLink = (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.velocity.enabled) ? "https://docs.papermc.io/velocity/security" : "http://www.spigotmc.org/wiki/firewall-guide/";
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index 356107688a5d40d1c462b164f61af82f4dfd3926..36f3913b80d9f2a8945bc7d1bc58a6cabe150338 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -304,6 +304,7 @@ public final class CraftServer implements Server {
+     private boolean overrideAllCommandBlockCommands = false;
+     public boolean ignoreVanillaPermissions = false;
+     private final List<CraftPlayer> playerView;
++    public boolean isReloading; // Paper
+     public int reloadCount;
+     private final io.papermc.paper.datapack.PaperDatapackManager datapackManager; // Paper
+     public static Exception excessiveVelEx; // Paper - Velocity warnings
+@@ -1009,6 +1010,7 @@ public final class CraftServer implements Server {
+ 
+     @Override
+     public void reload() {
++        this.isReloading = true; // Paper
```

Comment:
So this only covers 1 type of reload. There is the very much supported reload caused by `/minecraft:reload` which reloads datapacks, recipes, tags, advancements, etc. I think that should be covered by this as well.

Reply:
after thinking about it for some time i think instead of adding the isReloading API i should just add an internal boolean called isReloadingPlugins. and then move the check from the initServer method into the last if statement of the CraftServer#enablePlugins.